From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Zetarancio <zetarancio@users.noreply.github.com>
Date: Sun, 23 Feb 2026 00:00:00 +0000
Subject: [PATCH] dt-bindings: memory-controllers: add rockchip,rk3568-dmc

Document the devicetree binding for the RK3566/RK3568 Dynamic Memory
Controller (DMC) devfreq device.

DDR frequency scaling on these SoCs uses a shared-memory SIP protocol
with MCU-based frequency change completion, requiring a dedicated
completion interrupt.

Signed-off-by: Zetarancio <zetarancio@users.noreply.github.com>
---
 .../memory-controllers/rockchip,rk3568-dmc.yaml | 71 ++++++++++++++++
 1 file changed, 71 insertions(+)
 create mode 100644 Documentation/devicetree/bindings/memory-controllers/rockchip,rk3568-dmc.yaml

diff --git a/Documentation/devicetree/bindings/memory-controllers/rockchip,rk3568-dmc.yaml b/Documentation/devicetree/bindings/memory-controllers/rockchip,rk3568-dmc.yaml
new file mode 100644
index 000000000000..111111111111
--- /dev/null
+++ b/Documentation/devicetree/bindings/memory-controllers/rockchip,rk3568-dmc.yaml
@@ -0,0 +1,71 @@
+# SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause)
+# %YAML 1.2
+---
+$id: http://devicetree.org/schemas/memory-controllers/rockchip,rk3568-dmc.yaml#
+$schema: http://devicetree.org/meta-schemas/core.yaml#
+
+title: Rockchip RK3566/RK3568 DMC (Dynamic Memory Controller) device
+
+maintainers:
+  - Heiko Stuebner <heiko@sntech.de>
+
+description: |
+  The RK3566/RK3568 DMC devfreq device scales DDR frequency using a
+  shared-memory SIP protocol with the ARM Trusted Firmware.  Frequency
+  changes are performed by an on-chip MCU (DCF); the kernel is notified
+  of completion via a GIC interrupt.
+
+properties:
+  compatible:
+    enum:
+      - rockchip,rk3568-dmc
+
+  devfreq-events:
+    $ref: /schemas/types.yaml#/definitions/phandle
+    description:
+      Phandle to the DFI device that provides DDR utilization data.
+      Refer to Documentation/devicetree/bindings/devfreq/event/rockchip,dfi.yaml.
+
+  clocks:
+    maxItems: 1
+
+  clock-names:
+    items:
+      - const: dmc_clk
+
+  operating-points-v2: true
+
+  center-supply:
+    description:
+      Regulator supply for the memory controller logic voltage (vdd_logic).
+
+  interrupts:
+    maxItems: 1
+    description:
+      MCU DCF completion interrupt.  Fired when the on-chip MCU finishes
+      a DDR frequency change.
+
+  interrupt-names:
+    items:
+      - const: complete
+
+required:
+  - compatible
+  - devfreq-events
+  - clocks
+  - clock-names
+  - operating-points-v2
+  - center-supply
+  - interrupts
+  - interrupt-names
+
+additionalProperties: false
+
+examples:
+  - |
+    #include <dt-bindings/interrupt-controller/arm-gic.h>
+
+    dmc: dmc {
+        compatible = "rockchip,rk3568-dmc";
+        devfreq-events = <&dfi>;
+        clocks = <&scmi_clk 3>;
+        clock-names = "dmc_clk";
+        operating-points-v2 = <&dmc_opp_table>;
+        center-supply = <&vdd_logic>;
+        interrupts = <GIC_SPI 10 IRQ_TYPE_LEVEL_HIGH>;
+        interrupt-names = "complete";
+    };
-- 
2.34.1

